File: recursion/mergeSort.cpp
#include <bits/stdc++.h>
using namespace std;
void merge(int low, int mid, int high, vector<int> &arr) {
  int i = low;
  int j = mid + 1;
  vector<int> temp;
  while (i <= mid && j <= high) {
    if (arr[i] <= arr[j])
      temp.push_back(arr[i++]);
    else
      temp.push_back(arr[j++]);
  }
  while (i <= mid)
    temp.push_back(arr[i++]);
  while (j <= high)
    temp.push_back(arr[j++]);
  for (i = low; i <= high; i++) {
    arr[i] = temp[i - low];
  }
}
void mergeSort(int low, int high, vector<int> &arr) {
  if (low < high) {
    int mid = (low + high) / 2;
    mergeSort(low, mid, arr);
    mergeSort(mid + 1, high, arr);
    merge(low, mid, high, arr);
  }
}

int main() {
  vector<int> arr{1, 3, 5, 7, 9, 2, 4, 6, 8};
  mergeSort(0, arr.size() - 1, arr);
  for (auto it : arr)
    cout << it << " ";
  cout << endl;
}

File: sorting/MergeSort.py
def mergeSort(arr):
    n = len(arr)
    if n <= 1:
        return arr
    mid = n // 2
    left = arr[:mid]
    right = arr[mid:]
    leftSorted = mergeSort(left)
    rightSorted = mergeSort(right)
    return merge(leftSorted, rightSorted)


def merge(left, right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


# Example Usage
size = int(input("Enter the size of the Array:"))
num = []
for i in range(size):
    val = int(input())
    num.append(val)
print(f"Before Sorting : {num}")
n = len(num)
result = mergeSort(num)
print(f"After Sorting : {result}")

"""
Notes:
-Divide and conquer approach follows
-best for linked list bcoz no extra space required

stable : True 
In-place : False 

TC : o(nlogn)
    Best Case : O(nlogn) 
    Worst Case : O(nlogn) 
SC : O(n)
"""
